model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java

```diff
@@ -167,7 +167,7 @@ public void addFederatedIdentity(RealmModel realm, UserModel user, FederatedIden
         entity.setRealmId(realm.getId());
         entity.setIdentityProvider(identity.getIdentityProvider());
         entity.setUserId(identity.getUserId());
-        entity.setUserName(identity.getUserName().toLowerCase());
```

Comment:
Do we want a switch to enable/disable this behavior? This change can break existing deployments expecting the original username to be in lower-case.

Reply:
It is probably a good idea to do it, IMO.

Reply:
Can we just update this and document in the migration guide? Hopefully we can avoid (yet another) switch. But not 100% sure... If you guys rather think, we should keep backwards compatibility, I don't want to block this.

Reply:
I'll go then with the path of not adding any switch but updating both release and upgrade guides. I think most of the time, you really want to keep the original username. Also, I think one could always use a mapper to lowercase the value if they really need this data into tokens.

@thomasdarimont @sschu @tnorimat Based on your experience, do you have any strong opinions about this? Does it make sense to introduce a setting to keep storing the original username in lowercase (case-insensitive) when federating users from a broker?

Reply:
@pedroigor Hello, 
From my side, there is no strong opinion about it. IMO, it makes sense to me to add a switch make sense to keep backward compatibility.

Reply:
I also would try to not break compatibility, otherwise people will end up with an upper/lower case mixture. Maybe we could keep the old behaviour for existing realms and introduce the new behaviour for new realms only? One option to not use a switch here would be to edit the template for new realms to always use a username template mapper in its first broker login flow. We could document that in order to change the behaviour for existing realms, you can add a username template mapper.

Besides that, I think the search logic in the JPAUserProvider needs to be adapted since currently it always searches for lower username only, see https://github.com/keycloak/keycloak/blob/45e5e6cbbf966f31fb4b7d80a9a914b6731edc41/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L938 and https://github.com/keycloak/keycloak/blob/45e5e6cbbf966f31fb4b7d80a9a914b6731edc41/model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java#L994

Reply:
@sschu Those references are related to the querying users and not their federated identities. AFAIK, we don't have any query that uses https://github.com/keycloak/keycloak/blob/main/model/jpa/src/main/java/org/keycloak/models/jpa/entities/FederatedIdentityEntity.java#L64 in a predicate.

I like though your suggestion to keep the existing behavior for existing realms while introducing the new behavior only for new ones. I'm wondering if we should consider binding realms with a version so that we can change behavior like this one only for realms greater than or equal to the version where a realm was created?

Reply:
Ah, I assumed this was targeted at the username that Keycloak set's for a user based on the username from the external IDP. That's why I was also looking at the UsernameTemplateMapper. Binding switches to realm versions sounds like a nice idea to me. But. I would probably try to get more people's opinion on this since this could probably be used in several places.